refactor(parser): reduce rewind in checking if start of function type or constructor type

camchenry opened the pull request:
- part of https://github.com/oxc-project/oxc/issues/11334

Optimizes the early returns to do as much checking on the current token as possible before creating a checkpoint and doing lookahead. De-duplicates checkpoint creation in `lookahead` by inlining the methods so that only one checkpoint needs to be allocated.

camchenry commented:
Maybe a very slight perf win? Up to 1%, at best.
<img width="728" alt="Screenshot 2025-06-11 at 11 01 37 PM" src="https://github.com/user-attachments/assets/27dc60c9-e207-4b0a-85d5-87f4434f1c81" />

Boshen commented:
### Merge activity

* **Jun 12, 3:35 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

camchenry committed:
refactor(parser): reduce rewind in checking if start of function type or constructor type (#11622)

- part of https://github.com/oxc-project/oxc/issues/11334

Optimizes the early returns to do as much checking on the current token as possible before creating a checkpoint and doing lookahead. De-duplicates checkpoint creation in `lookahead` by inlining the methods so that only one checkpoint needs to be allocated.
